controller/konnect/reconciler_credential_secrets.go

```diff
@@ -0,0 +1,418 @@
+					secretIsUsedByConsumerAttachedToKonnectControlPlane(mgr.GetClient()),
+				),
+			),
+		).
+		Watches(&configurationv1.KongConsumer{},
+			handler.EnqueueRequestsFromMapFunc(enqueueSecretForKongConsumer),
+			builder.WithPredicates(
+				predicate.NewPredicateFuncs(objRefersToKonnectGatewayControlPlane[configurationv1.KongConsumer]),
+			),
+		).
+		// NOTE: We use MatchEveryOwner because we set both the KongConsumer and
+		// the Secret holding the credentials as owners.
+		// The KongConumser is set for obvious reasons, when that's deleted we want
+		// to delete the managed credential to be deleted as well.
+		// The Secret is set as owner because we want to trigger the reconciliation
+		// of the secret (this controller) so that the managed credential is enforced
+		// (e.g. recreated).
+		Owns(&configurationv1alpha1.KongCredentialBasicAuth{}, builder.MatchEveryOwner).
+		// TODO: add more credentials types support.
+		Complete(r)
+}
+
+func enqueueSecretForKongConsumer(ctx context.Context, obj client.Object) []reconcile.Request {
+	consumer, ok := obj.(*configurationv1.KongConsumer)
+	if !ok {
+		return nil
+	}
+
+	ret := make([]ctrl.Request, 0, len(consumer.Credentials))
+	for _, secretName := range consumer.Credentials {
+		ret = append(ret, ctrl.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      secretName,
+				Namespace: consumer.Namespace,
+			},
+		})
+	}
+	return ret
+}
+
+// secretIsUsedByConsumerAttachedToKonnectControlPlane returns true if the Secret
+// is used by KongConsumer which refers to a KonnectGatewayControlPlane.
+func secretIsUsedByConsumerAttachedToKonnectControlPlane(cl client.Client) func(obj client.Object) bool {
+	return func(obj client.Object) bool {
+		secret, ok := obj.(*corev1.Secret)
+		if !ok {
+			ctrllog.FromContext(context.Background()).Error(
+				operatorerrors.ErrUnexpectedObject,
+				"failed to run predicate function",
+				"expected", "Secret", "found", reflect.TypeOf(obj),
+			)
+			return false
+		}
+
+		// List consumers using this Secret as credential.
+		kongConsumerList := configurationv1.KongConsumerList{}
+		err := cl.List(
+			context.Background(),
+			&kongConsumerList,
+			client.MatchingFields{
+				IndexFieldKongConsumerReferencesSecrets: secret.GetName(),
+			},
+		)
+		if err != nil {
+			return false
+		}
+
+		for _, kongConsumer := range kongConsumerList.Items {
+			cpRef := kongConsumer.Spec.ControlPlaneRef
+			if cpRef != nil &&
+				cpRef.Type == configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+// Reconcile reconciles a Secrets that are used as Consumers credentials.
+func (r *KongCredentialSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var secret corev1.Secret
+	if err := r.client.Get(ctx, req.NamespacedName, &secret); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	const entityTypeName = "Secret"
```

Comment:
This is used in row 151 only as far as I see. Can we just hardcode it there? It doesn't seem something likely to change or to be reused somewhere else

Reply:
23f82e0d384b1b65e3ba166edb623272942c7636